File: FactoryMethod/factory_method.kt
// Animal 인터페이스
interface Animal {
    fun speak(): String
}

// Dog 클래스
class Dog : Animal {
    override fun speak(): String = "Woof!"
}

// Cat 클래스
class Cat : Animal {
    override fun speak(): String = "Meow!"
}

// AnimalFactory 클래스
class AnimalFactory {
    fun createAnimal(type: String): Animal? {
        return when (type.lowercase()) {
            "dog" -> Dog()
            "cat" -> Cat()
            else -> null
        }
    }
}

// 실행 메인 함수
fun main() {
    val factory = AnimalFactory()

    val animal1 = factory.createAnimal("dog")
    val animal2 = factory.createAnimal("cat")

    println(animal1?.speak() ?: "Unknown animal")
    println(animal2?.speak() ?: "Unknown animal")
}

File: FactoryMethod/factory_method.cpp
#include <iostream>
#include <memory>
#include <string>

using namespace std;

// 추상 클래스 Animal
class Animal {
public:
    virtual string speak() const = 0;
    virtual ~Animal() {}
};

// Dog 클래스
class Dog : public Animal {
public:
    string speak() const override {
        return "Woof!";
    }
};

// Cat 클래스
class Cat : public Animal {
public:
    string speak() const override {
        return "Meow!";
    }
};

// AnimalFactory 클래스
class AnimalFactory {
public:
    unique_ptr<Animal> createAnimal(const string& type) {
        if (type == "dog") {
            return make_unique<Dog>();
        } else if (type == "cat") {
            return make_unique<Cat>();
        } else {
            return nullptr;
        }
    }
};

int main() {
    AnimalFactory factory;

    auto dog = factory.createAnimal("dog");
    auto cat = factory.createAnimal("cat");

    if (dog) {
        cout << "Dog says: " << dog->speak() << endl;
    }

    if (cat) {
        cout << "Cat says: " << cat->speak() << endl;
    }

    return 0;
}
